### ISSUE TITLE
Help Wanted, Possible Memory leak?

### ISSUE THREAD

**Issue by Starystars67** [mde/ejs#392]
Hello, okay so i know there have been other issues on the same subject but im at a loss with this and need some help please. Okay so why do i think EJS is the cause? here is some screen shots of heap snapshots in comparison between start up and after autocannon. 
![image](https://user-images.githubusercontent.com/15389482/43230364-d64b2408-905f-11e8-82f4-ab631ec3622d.png)
As you can see in the above image there is a sizable difference in file size given its a web server.
![image](https://user-images.githubusercontent.com/15389482/43230227-609f8f28-905f-11e8-9dc3-9cd324e202d2.png)

so after taking this and searching through the code I boiled it down to line 573 in ejs.js which is located in the lib folder of the ejs library. 
Further scrolling through the leaked strings I found this which i think confirms the leak to be related to EJS: 
![image](https://user-images.githubusercontent.com/15389482/43230296-a3041866-905f-11e8-84c4-6011a1ac7e44.png)

I will add more as I get it, Again any help is very much appreciated and can someone please add the help-wanted tag.

Many Thanks.

EDIT: Code!! Because everyone loves code.
```
const memwatch   = require('memwatch-next');
var moment = require('moment');

// load the things we need
var express = require('express');
var app = express();

// set the view engine to ejs
app.set('view engine', 'ejs');

// use res.render to load up an ejs view file

// index page
app.get('/', function(req, res) {
	res.render('public/landingpage', {moment: moment});
});

app.listen(8080);
console.log('8080 is the magic port');

memwatch.on('leak', (info) => {
  console.log(info)
});
```

**Comment by m-majetic** [comment]
Greetings !
I have opened an issue as well, regarding a memory leak.
Did you resolve the issue or does it still persist ?

**Comment by m-majetic** [comment]
I see that other issues haven't been closed as well. 
It makes me question, is everything is alright with the development of EJS ?

**Comment by Starystars67** [comment]
Hmm makes you wonder doesn't it.

**Comment by mde** [comment]
I've taken a bit of time to look into this. From what I can tell from your screenshots, and a brief session with the Node debugger myself, it looks like those strings are the source strings for the constructed function that the Template objects use to execute and render templates here: https://github.com/mde/ejs/blob/master/lib/ejs.js#L626 This is a result of using the `Function` constructor, which is a built-in.

I took Express out of the equation completely, just loading EJS into a Node script and calling `render` with template text of empty string in a tight loop, using `setImmediate` to ensure GC can occur. Both with requested GC (`global.gc()`) and normal, Node-scheduled GC, I did see these strings getting GC'd. The behavior was almost the same with an inline use of `Function` constructor with a dummy function body.

I'm not sure if these strings are slow to GC, or what, but I have not been able to isolate anything specific in EJS that looks anything like a memory leak. I do see memory go up, but the behavior looks almost identical with the inline use of a Function constructor, or even with a complete no-op in that same loop.

Having said that, I will continue to dig into this and see if I can track down something more concrete. I only took a cursory look at this today, so I may very well be missing something else.

Thanks for your patience with this.
